x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts

```diff
@@ -83,3 +83,32 @@ export const PrebuiltRuleAsset = BaseCreateProps.omit(BASE_PROPS_REMOVED_FROM_PR
       version: RuleVersion,
     })
   );
+
+/**
+ * Creates a Map of the fields that are upgradable during the Upgrade workflow, by type.
+ * Creating this Map dynamically, based on BaseCreateProps and TypeSpecificFields, ensures that we don't need to:
+ *  - manually add rule types to this Map if they are created
+ *  - manually add or remove any fields if they are added or removed to a specific rule type
+ *  - manually add or remove any fields if we decide that they should not be part of the upgradable fields.
+ */
+function createUpgradableRuleFieldsByTypeMap() {
+  const baseFields = Object.keys(
+    BaseCreateProps.omit(BASE_PROPS_REMOVED_FROM_PREBUILT_RULE_ASSET).shape
+  );
+
+  // Since we used Zod's .transform() method over TypeSpecificCreateProps above,
+  // the type of TypeSpecificFields is actually z.ZodEffects, and therefore does
+  // not have access to an options property. We need to transform it back into
+  // a z.ZodDiscriminatedUnion, accessing its underlying schema.
+  const TypeSpecificFieldsSchema = TypeSpecificFields._def.schema;
```

Comment:
If I understand correctly, we have a zod transform that removes `response_actions` from certain rule types. Therefore, I expect `response_actions` not to be part of `UPGRADABLE_RULES_FIELDS_BY_TYPE_MAP`. However, `_def.schema` actually accesses the initial definitions where `response_actions` are included so they become upgradeable fields.

Reply:
I think what you need is to use omit instead of relying on transform:

```ts
export type TypeSpecificFields = z.infer<typeof TypeSpecificFields>;
export const TypeSpecificFields = z.discriminatedUnion('type', [
  EqlRuleCreateFields,
  QueryRuleCreateFields.omit({ response_actions: true }),
  SavedQueryRuleCreateFields.omit({ response_actions: true }),
  ThresholdRuleCreateFields,
  ThreatMatchRuleCreateFields,
  MachineLearningRuleCreateFields,
  NewTermsRuleCreateFields,
  EsqlRuleCreateFields,
]);
```

Reply:
> I expect response_actions not to be part of UPGRADABLE_RULES_FIELDS_BY_TYPE_MAP

Yes, thanks for that catch. It indeed includes `response_actions`, which is exactly what I'm trying to prevent.

The solution you proposed above would be the easiest way forward, and would solve the issue with the `response_action` field being present in the Map, but the reason we initially went with the more complex Zod `.transform()` is because we discussed with @banderror that we want to avoid having and _additional_ place where the rule types are listed, and which needs to be modified if a new rule type is added.

I think it makes sense: when ES|QL was added, and the Detection Engine team added the functionality for a corresponding security rule, but we did not go through our codebase to check where rule types where listed and we needed to take some action.

So, yeah, not sure. I'll try to find another way of generating the Map without the `response_action` field using the output of the transform. 

One possibility is just filtering out unwanted `fields` from the type specific fields within the function that creates the map. Something like:

Reply:
@xcrzx and FYI @banderror 

OK, the apporach I proposed above is technically feasible, and we are still able to generate the derived types without listing all the rule types.

I pushed the changes. Take a look and let me know what you think.

Reply:
It looks like the transformation now turns all type-specific props into `Record<string, any>`:

![image](https://github.com/user-attachments/assets/7843db47-f857-4260-8961-5551903806a7)

Reply:
> I think it makes sense: when ES|QL was added, and the Detection Engine team added the functionality for a corresponding security rule, but we did not go through our codebase to check where rule types where listed and if we needed to take some action.

Can we use the power of Typescript to ensure those types are always in sync?

```ts
import type { IsEqual } from 'type-fest';
// Make sure the type-specific fields contain all the same rule types as the type-specific rule params
// Here will be a type error if the types are not equal
const areTypesEqual: IsEqual<
  typeof TypeSpecificRuleParams._type.type,
  typeof TypeSpecificFields._type.type
> = true;
```

Reply:
Another option, just use tests:

```ts
expect(TypeSpecificRuleParams._type.type).toEqual(TypeSpecificFields._type.type);
```